### PR TITLE
Packaging cleanup, migration to full setuptools without distutils

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Description

Move from distutil to setuptools. (Second step toward pyproject.toml after #4295 ).

We might also want to remove ``doc/`` and ``tests/`` directory from the packaging ? Are you using the tests when packaging for debian @sandrotosi ?

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :hammer: Refactoring  |

## Related Issue

Relates to #4193

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,164 +1,89 @@
...
+        install_requires=[
+            "astroid>=2.5.2,<2.7",
+            "isort>=4.2.5,<6",
+            "mccabe>=0.6,<0.7",
+            "toml>=0.7.1",
+        ],
+        extras_require={
+            ':sys_platform=="win32"': ["colorama"],
+            "docs": doc_extra_requires,
+            "dev": dev_extra_requires,
```
Comment (cdce8p) [inline setup.py:77]: I don't really see a need to add `dev` as an `extras_require` here. The usual way  should be to fork and clone the repo and then do
```bash
cd pylint
pip install -e .
pip install -r requirements_test.txt
```

**Review by cdce8p — CHANGES_REQUESTED** [review]
I would prefer if the metadata where to be defined in `setup.cfg` instead. See here, for reference: https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html
It might also make sense to separate the build system changes from the removal of `bin`, `debian`, and `man` and put that in another MR.

File: setup.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply setup.py:77]: I like being able to install everything at once with ``pip3 install -e .[dev]``, would you mind if we keep it ?

**Comment by Pierre-Sassoulas** [comment]
I opened another MR to separate removal of legacy build file and setuptools migration. Regarding the move to ``setup.cfg`` I think it's going to be difficult because we have the version defined inside the package and also the test_requires and extra_requires so we need to get the value with code.

File: setup.py (hunk 1 above)
Comment (cdce8p) [reply setup.py]: I understand your reasoning. However in that case we should probably loosen the test requirements inside `setup.py` and `setup.cfg`. `install_requires` and `extras_require` should usually **not** pin exact versions. From that viewpoint, I would still recommend installing the test requirements via `pip install -r requirements_test.txt`.

**Comment by Pierre-Sassoulas** [comment]
Nevermind that, most of it is doable with setup.cfg. For version we have ``pbr`` and ``setuptools_scm`` that could work. I don't know if we could access the version inside pylint for ``pylint --version`` though. But the setup.py became very reasonable already.

File: setup.cfg [hunk 2]
```diff
@@ -1,6 +1,85 @@
-[bdist_rpm]
-packager = Sylvain Thenault  <sylvain.thenault@logilab.fr>
-provides = pylint
+[metadata]
+name = pylint
+license = GPL-2.0-or-later
+license_file = COPYING
```
Comment (cdce8p) [inline setup.cfg:4]: `license_file` is deprecated and has been replace by `license_files`.
I would also suggest to add `CONTRIBUTORS.txt` here as well. What do you think?

File: setup.cfg [hunk 3]
```diff
@@ -1,6 +1,85 @@
...
+description = python code static checker
+keywords = static code analysis linter python lint
+long_description = file: README.rst
+long_description_content_type = text/x-rst; charset=UTF-8
```
Comment (cdce8p) [inline setup.cfg:8]:
```suggestion
long_description_content_type = text/x-rst
```
`UTF-8` is already the default if omitted. https://packaging.python.org/specifications/core-metadata/#description-content-type

File: setup.cfg [hunk 4]
```diff
@@ -1,6 +1,85 @@
...
+    Topic :: Software Development :: Debuggers
+    Topic :: Software Development :: Quality Assurance
+    Topic :: Software Development :: Testing
+
+[options]
+zip_safe = True
+packages = find:
+python_requires = ~=3.6
+py_modules = easy_install
+setup_requires =
```
Comment (cdce8p) [inline setup.cfg:42]: If unused, it can probably be removed.

File: setup.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply setup.py]: I kept the old structure (I just removed astroid from test dependencies as it's a dependency already) and put it in the setup.cfg. Less files to keep track of I think. See https://github.com/PyCQA/pylint/pull/4294/commits/1291354f35730df4411249a7e544f6cf4dc118fe

File: setup.py [hunk 5]
```diff
@@ -1,166 +1,16 @@
...
-        setup_requires=pytest_runner,
-        tests_require=["pytest", "pytest-benchmark"],
-        project_urls=project_urls,
-        **kwargs,
-    )
-
+with open(HERE / "pylint/__pkginfo__.py", encoding="UTF-8") as f:
+    exec(f.read(), __pkginfo__)  # pylint: disable=exec-used
 
 if __name__ == "__main__":
```
Comment (cdce8p) [inline setup.py:15]: I don't think its necessary here. `setup.py` is only ever executed to build the packages.
The default I'm used to is
```py
from setuptools import setup

setup(
  ...
)
```

File: tox.ini [hunk 6]
```diff
@@ -6,25 +6,25 @@ skip_missing_interpreters = true
 
 [testenv:pylint]
 deps =
-    -r {toxinidir}/requirements_test_min.txt
-    pre-commit==2.10.1
+    -e {toxinidir}[test_minimal]
```
Comment (cdce8p) [inline tox.ini:9]: IMO `tox` should always use the pinned requirements files. As explained earlier, `extras_require` should not pin exact versions.

File: tox.ini [hunk 7]
```diff
@@ -60,32 +60,28 @@ commands =
...
-        --benchmark-only \
-        --benchmark-save=batch_files \
-        --benchmark-save-data \
-        --benchmark-autosave {toxinidir}/tests \
-        --benchmark-group-by="group" \
-        {posargs:}
+    pytest --exitfirst --failed-first --benchmark-only \
+    --benchmark-save=batch_files --benchmark-save-data \
+    --benchmark-autosave {toxinidir}/tests --benchmark-group-by="group" \
+    {posargs:}
```
Comment (cdce8p) [inline tox.ini:78]: Is that really an improvement?

**Review by cdce8p — CHANGES_REQUESTED** [review]
Left some comments. Even if you add an `extras_require`, please don't remove the `requirements_test.txt` files. They are need to pin exact versions for CI testing.

File: setup.py (hunk 5 above)
Comment (Pierre-Sassoulas) [reply setup.py:15]: A test for python 3.8 and 3.9 was failing without it (benchmark I think)

File: tox.ini (hunk 7 above)
Comment (Pierre-Sassoulas) [reply tox.ini:78]: Fair enough 😄

File: setup.cfg (hunk 3 above)
Comment (Pierre-Sassoulas) [reply setup.cfg:8]: 👍

File: setup.py (hunk 5 above)
Comment (cdce8p) [reply setup.py:15]: Probably a bad test then 😄

File: MANIFEST.in [hunk 8]
```diff
@@ -2,10 +2,10 @@ include ChangeLog
 include CONTRIBUTORS.txt
 include COPYING
 include README.rst
-include bin/*
-include examples/*.py examples/pylintrc examples/pylintrc_camelcase
 include elisp/startup elisp/*.el
 include pytest.ini
-recursive-include doc *.bat *.png *.py *.svg *.rst *.txt Makefile
+recursive-include doc *.bat *.png *.py *.svg *.rst *.txt Makefile .md
 graft tests
```
Comment (cdce8p) [inline MANIFEST.in:8]: You probably want to remove this line here: `graft tests`. It will include all files in the `tests` folder recursively.

File: setup.cfg (hunk 4 above)
Comment (Pierre-Sassoulas) [reply setup.cfg:42]: I'm planning on maybe using ``pbr`` or ``setuptools_scm`` or something else (?) to handle the version. Do you use anything like that yourself ? Seems like it could remove a whole lot of manipulation of the version during release too...

File: setup.cfg [hunk 9]
```diff
@@ -1,3 +1,86 @@
...
+license_file = COPYING
+description = python code static checker
+keywords = static code analysis linter python lint
+long_description = file: README.rst
+long_description_content_type = text/x-rst; charset=UTF-8
+author = Python Code Quality Authority
+author_email = code-quality@python.org
+url = https://github.com/PyCQA/pylint
+project_urls =
+    What's New=https://pylint.pycqa.org/en/latest/whatsnew/
```
Comment (cdce8p) [inline setup.cfg:15]: Although not technically needed, can you add quotes here? That would improve the readability and enable better syntax highlighting.
```ini
project_urls =
    "What's New" = https://....
```

File: setup.cfg (hunk 2 above)
Comment (Pierre-Sassoulas) [reply setup.cfg]: 👍

File: setup.cfg [hunk 10]
```diff
@@ -1,3 +1,88 @@
...
+    Programming Language :: Python :: 3 :: Only
+    Programming Language :: Python :: Implementation :: CPython
+    Programming Language :: Python :: Implementation :: PyPy
+    Topic :: Software Development :: Debuggers
+    Topic :: Software Development :: Quality Assurance
+    Topic :: Software Development :: Testing
+
+[options]
+zip_safe = True
+packages = find:
```
Comment (cdce8p) [inline setup.cfg:41]: To not include just all packages, you could add a new section:
```ini
[options.packages.find]
include =
    pylint*
```

File: MANIFEST.in (hunk 8 above)
Comment (Pierre-Sassoulas) [reply MANIFEST.in:8]: Yeah, I was wondering if we keep the doc folder and the test folder in the packaged result at all... (We exclude .pyc file from tests directory on the next line).

File: setup.cfg (hunk 4 above)
Comment (cdce8p) [reply setup.cfg:44]: I haven't used any of them so I can't give you an advice there unfortunately.

File: setup.py (hunk 5 above)
Comment (Pierre-Sassoulas) [reply setup.py:15]: It's ``test_baseline_benchmark_j1_all_checks_lots_of_files``, I have no idea why it fails though. 
```python
        try:
            ok = dist.parse_command_line()
        except DistutilsArgError as msg:
>           raise SystemExit(gen_usage(dist.script_name) + "\nerror: %s" % msg)
E           SystemExit: usage: __main__.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
E              or: __main__.py --help [cmd1 cmd2 ...]
E              or: __main__.py --help-commands
E              or: __main__.py cmd --help
E           
E           error: no commands supplied
```

**Comment by Pierre-Sassoulas** [comment]
> Please don't remove the `requirements_test.txt` files. They are need to pin exact versions for CI testing.

I thought the the ``setup.cfg`` could also work for that goal ? Are we talking about the hash part ?

File: MANIFEST.in (hunk 8 above)
Comment (cdce8p) [reply MANIFEST.in:8]: I would probably remove most lines in the `MANIFEST.in`. It's only relevant for the `sdist` to begin with and there isn't a good reason to keep them around if they aren't used.

```
include ChangeLog             # not used
include CONTRIBUTORS.txt      # included through `license_files`
include COPYING               # included through `license_files`
include README.rst            # included `EGG-INFO` / `PKG-INFO`
include elisp/startup elisp/*.el  # ?
include pytest.ini            # not necessary
recursive-include doc *.bat *.png *.py *.svg *.rst *.txt Makefile .md  # Unused
recursive-exclude tests *.pyc .benchmark  # Can be removed if `[options.packages.find]` is added
recursive-exclude script *    # Same here
recursive-exclude examples *  # and here
```

**Comment by cdce8p** [comment]
> > Please don't remove the `requirements_test.txt` files. They are need to pin exact versions for CI testing.
> 
> I thought the the `setup.cfg` could also work for that goal ? Are we talking about the hash part ?

Unfortunately `extras_require` doesn't work and shouldn't be used for that purpose. Tbh I have only seen `install -e .[test]` as part of legacy configs that haven't been updated yet.
The reason is, that users would expect to be able to install the `extra` seamlessly in their environment. That is only possible if the requirements aren't fixed as it otherwise might create dependency resolution conflicts. The idea behind the `requirements` file on the other hand is exactly to provide a fixed environment that can be used for testing and will always yield the same result.

File: setup.cfg (hunk 4 above)
Comment (Pierre-Sassoulas) [reply setup.cfg]: ``version = attr: pylint.__version__`` make a temporary solution. Might change that later but we'd have to change the tag from ``pylint-x.y.z`` to ``x.y.z`` or ``vx.y.z``

Comment (sandrotosi) [reply setup.cfg]: > I'm planning on maybe using `pbr` or `setuptools_scm` or something else (?) to handle the version. Do you use anything like that yourself ? Seems like it could remove a whole lot of manipulation of the version during release too...

from a packager perspective, pbr and setuptools_scm are pretty annoying :) if you just need to read the version of pylint from withing setup.py you can choose some of the options available here: https://packaging.python.org/guides/single-sourcing-package-version/

File: setup.cfg [hunk 11]
```diff
@@ -1,3 +1,91 @@
...
+    CONTRIBUTORS.txt
+description = python code static checker
+keywords = static code analysis linter python lint
+long_description = file: README.rst
+long_description_content_type = text/x-rst
+author = Python Code Quality Authority
+author_email = code-quality@python.org
+url = https://github.com/PyCQA/pylint
+project_urls =
+    "What's New"=https://pylint.pycqa.org/en/latest/whatsnew/
```
Comment (cdce8p) [inline setup.cfg:16]: Just for looks, I would add some whitespaces around `=`

**Comment by sandrotosi** [comment]
> We might also want to remove `doc/` and `tests/` directory from the packaging ? Are you using the tests when packaging for debian @sandrotosi ?

we do run tests and build the documentation during debian package build, but dont worry about that: i've been gradually migrating all my packages from fetching new upstream releases from PyPI to GitHub releases: it's just easier downstream (we get all the files that we need) and you can keep the PyPI release small (without extra files end users dont really need when pip installing pylint) -- thanks for asking tho!

File: setup.cfg [hunk 12]
```diff
@@ -1,3 +1,91 @@
...
+    Programming Language :: Python :: 3.9
+    Programming Language :: Python :: 3 :: Only
+    Programming Language :: Python :: Implementation :: CPython
+    Programming Language :: Python :: Implementation :: PyPy
+    Topic :: Software Development :: Debuggers
+    Topic :: Software Development :: Quality Assurance
+    Topic :: Software Development :: Testing
+[options.packages.find]
+include =
+    pylint*
```
Comment (cdce8p) [inline setup.cfg:42]: Maybe move this section after `[options]`?
Btw, I believe you still need `packages = find:` in the `[options]` section.

**Comment by Pierre-Sassoulas** [comment]
>  i've been gradually migrating all my packages from fetching new upstream releases from PyPI to GitHub releases: it's just easier downstream (we get all the files that we need) and you can keep the PyPI release small (without extra files end users dont really need when pip installing pylint) 

Amazing, we're going to be able to make the package a lot smaller !

File: MANIFEST.in (hunk 8 above)
Comment (Pierre-Sassoulas) [reply MANIFEST.in:8]: Yeap, I was waiting for sandrotosi's opinion for that :)

**Comment by cdce8p** [comment]
@Pierre-Sassoulas Regarding the failing benchmark test. The reason is that the test tries to import all modules in the root dir to check for a `register` function. A `patch` fixes that:
```py
from unittest.mock import patch

# ....

    def test_baseline_benchmark_j1_all_checks_lots_of_files(self, benchmark):
        # ...
        with patch("os.listdir", return_value=["pylint", "tests"]):
            register_plugins(
                linter, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
            )
        # ...
```

**Comment by cdce8p** [comment]
You should also be able to include `.isort.cfg` and `pytest.ini` in `setup.cfg`.

File: tox.ini (hunk 6 above)
Comment (Pierre-Sassoulas) [reply tox.ini:9]: I'm basing this practice upon this stackoverflow answer : https://stackoverflow.com/a/28842733/2519059

I think extra_require are not always expected to be installable in alongside other packages. Especially if the extra require is called dev, test or doc.

Comment (cdce8p) [reply tox.ini:9]: I believe this is outdated, as it's from 2016. Best practices for python packaging are constantly changing if existing at all. They also strongly depend on who you ask. Just take a look at the official packaging guide, it mentions `requirements` files: https://packaging.python.org/discussions/install-requires-vs-requirements/#requirements-files
Even that is probably outdated by now and AFAIK the current standard would be to use `requirements.in` files with `pip-compile`.

In conclusion, I think the current system works really well. Not sure it's really worth changing, just to be able to do `pip install -e .[tests]` when `pip install -r requirements_tests.txt` works as well.

Comment (Pierre-Sassoulas) [reply tox.ini:9]: To be fair this is only temporary as if we move to ``pyproject.toml``and poetry which is the next step in my opinion, the dev requirement would be under ``[tool.poetry.dev-dependencies]`` (https://python-poetry.org/docs/pyproject/#dependencies-and-dev-dependencies).

**Review by hippo91 — APPROVED** [review]
@Pierre-Sassoulas great work! Unfortunately i'm not familiar with those packaging tools so i'm afraid i won't be able to give good advice here.

File: .github/workflows/ci.yaml [hunk 13]
```diff
@@ -32,7 +32,7 @@ jobs:
         id: generate-python-key
         run: >-
           echo "::set-output name=key::base-venv-${{ env.CACHE_VERSION }}-${{
-            hashFiles('pylint/__pkg_info__.py', 'requirements_test.txt',
+            hashFiles('pylint/__pkg_info__.py', 'requirements_test.txt', 'setup.cfg',
```
Comment (cdce8p) [inline .github/workflows/ci.yaml:35]: Since `pylint/__pkg_info__.py` doesn't contain the requirements anymore, it isn't necessary and can be removed here. Just so it's the same for all `hashFiles` calls, would you mind putting `setup.cfg` at the first position (the one `__pkg_info__.py` has right now)?

File: MANIFEST.in [hunk 14]
```diff
@@ -1,11 +1,6 @@
...
-include pytest.ini
-recursive-include doc *.bat *.png *.py *.svg *.rst *.txt Makefile
-graft tests
-recursive-exclude tests *.pyc
+exclude ChangeLog # At least 'https://pylint.pycqa.org/en/latest/whatsnew/' is not downloaded for each install
+recursive-exclude elisp *
+recursive-exclude doc *
+recursive-exclude tests *
+recursive-exclude script *
+recursive-exclude examples *
```
Comment (cdce8p) [inline MANIFEST.in:6]: These lines should not be necessary.

File: requirements_test_min.txt [hunk 15]
```diff
@@ -1,3 +1,2 @@
-astroid==2.5.2
```
Comment (cdce8p) [inline requirements_test_min.txt:1]: I might have missed it, but why remove `astroid` here? The idea as I recall was to pin it for the CI tests.

File: setup.cfg [hunk 16]
```diff
@@ -1,5 +1,77 @@
...
+    "What's New" = https://pylint.pycqa.org/en/latest/whatsnew/
+    "Bug tracker" = https://github.com/PyCQA/pylint/issues
+    "Discord server" = https://discord.gg/kFebW799
+    "User manual" = http://pylint.pycqa.org/en/latest/
+    "Contributing" = http://pylint.pycqa.org/en/latest/development_guide/contribute.html
+    "Technical references" = http://pylint.pycqa.org/en/latest/technical_reference/index.html
+    "Mailing list" = mailto:code-quality@python.org
+
+[options]
+packages = find:
```
Comment (cdce8p) [inline setup.cfg:43]: I think, during the rebase, you removed the `[options.packages.find]` section. Or is there another reason?
```ini
[options.packages.find]
include =
    pylint*
```

**Review by cdce8p — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml (hunk 13 above)
Comment (Pierre-Sassoulas) [reply .github/workflows/ci.yaml:35]: There's still the version in ``__pkg_info__.py``, or was it just for the requirements ?

File: requirements_test_min.txt (hunk 15 above)
Comment (Pierre-Sassoulas) [reply requirements_test_min.txt:1]: Astroid is already defined as a "real" dependency in install_requires.

File: setup.cfg (hunk 16 above)
Comment (Pierre-Sassoulas) [reply setup.cfg:43]: I did not manage to make the exclude works when using this, that's why I added the Manifest.in. Might try harder if you think it's a problem ? :D

File: requirements_test_min.txt (hunk 15 above)
Comment (cdce8p) [reply requirements_test_min.txt:1]: Yeah, but it's only loosely pinned there: `astroid>=2.5.2,<2.7`. That's why I added it to the `requirements` file originally.

File: .github/workflows/ci.yaml (hunk 13 above)
Comment (cdce8p) [reply .github/workflows/ci.yaml:35]: It was just for the requirements. So should be safe to remove.

File: setup.cfg (hunk 16 above)
Comment (cdce8p) [reply setup.cfg:43]: That's probably a caching issue. Try deleting the `pylint.egg-info`, `build`, and `dist` directories in between `python setup.py sdist bdist_wheel` runs.

File: requirements_test_min.txt (hunk 15 above)
Comment (Pierre-Sassoulas) [reply requirements_test_min.txt:1]: OK, it makes sense.

File: setup.cfg (hunk 16 above)
Comment (Pierre-Sassoulas) [reply setup.cfg:43]: Ok, I'll try that. Could you review : #4298, I would want to merge it before this one :)

Comment (Pierre-Sassoulas) [reply setup.cfg:43]: Probably the cache as you said 😄

File: requirements_test_min.txt [hunk 17]
```diff
@@ -1,3 +1,3 @@
...
+astroid==2.5.2 # Pinned to a specific version for tests
```
Comment (cdce8p) [inline requirements_test_min.txt:1]:
```suggestion
astroid==2.5.2  # Pinned to a specific version for tests
```

File: setup.cfg [hunk 18]
```diff
@@ -1,5 +1,81 @@
...
+install_requires =
+    astroid>=2.5.2,<2.7
+    isort>=4.2.5,<6
+    mccabe>=0.6,<0.7
+    toml>=0.7.1
+    colorama;sys_platform=="win32"
+python_requires = ~=3.6
+zip_safe = True
```
Comment (cdce8p) [inline setup.cfg:51]: I'm not really sure how this settings works, which is why I tend to avoid it and just leave it out. AFAIK `setuptools` can automatically detect if the package is `zip_safe`. Thus this wouldn't be needed necessarily.

**Review by cdce8p — APPROVED** [review]
Looks good overall! Thanks for your work @Pierre-Sassoulas
I left two small last comments. After they are addressed, this MR is ready to go.

File: setup.cfg (hunk 18 above)
Comment (Pierre-Sassoulas) [reply setup.cfg:51]: setuptools is conservative before saying a package is zip_safe, that's why you could want to specify it. pylint performance problem probably do not stem from not being zipped though.

Comment (cdce8p) [reply setup.cfg:51]: 👌🏻

Comment (Pierre-Sassoulas) [reply setup.cfg]: Apparently it's an old option that is not relevant for pip but was used for eggs. (So I removed it) See https://github.com/python/mypy/issues/8802#issuecomment-628249236
